src/mb_enhanced_cover_art_uploads/providers/base.ts

```diff
@@ -194,3 +180,129 @@ export abstract class HeadMetaPropertyProvider extends CoverArtProvider {
+            if (sameImages.length <= 1) {
+                results.push(sameImages[0][1]);
+                continue;
+            }
+
+            // If one of them is the front cover, we can just skip all track images
+            const frontCover = sameImages.find((pair) => pair[0].types?.includes(ArtworkTypeIDs.Front));
+            if (frontCover) {
+                results.push(frontCover[1]);
+                continue;
+            }
+
+            // Need to merge the images. We still have access to all track
+            // numbers, so we can create a new comment.
+            const allTrackNumbers = sameImages
+                .map((pair) => pair[0].comment?.match(/Tracks? ([\d\s,]+)/)?.[1] ?? '')
+                .flatMap((commentedTrackNos) => commentedTrackNos.split(',').map((num) => num.trim()))
+                .filter((num) => !!num);
+            const newComment = this.#createTrackImageComment(allTrackNumbers);
+            const mainImage = sameImages[0][1];
+            results.push({
+                ...mainImage,
+                comment: mainImage.comment?.replace(/Tracks? [\d\s,]*\d/, newComment) ?? newComment,
+            });
+        }
+
+        return Promise.resolve(results);
+    }
+
+    protected imageToThumbnailUrl(imageUrl: string): string {
+        // To be overridden by subclass if necessary.
+        return imageUrl;
+    }
+
+    protected async mergeTrackImages(trackImages: Array<ParsedTrackImage | undefined>, mainUrl: string | undefined, byContent: boolean): Promise<CoverArt[]> {
+        const allTrackImages = filterNonNull(trackImages);
+        // First pass: URL only
+        const groupedImages = this.#groupIdenticalImages(allTrackImages, mainUrl);
+
+        if (groupedImages.size > 1 && byContent) {
+            // Second pass: Thumbnail content
+            LOGGER.info('Deduplicating track images by content, this may take a while…');
+            const mainDigest = mainUrl ? await this.#urlToDigest(mainUrl) : /* istanbul ignore next: Difficult to cover */ '';
+            const dataToOriginal: Map<string, string> = new Map();
+            // Convert all track URLs to digests describing their content.
+            const trackDigests = await Promise.all([...groupedImages.entries()]
+                .map(async ([coverUrl, trackCovers]) => {
+                    const digest = await this.#urlToDigest(coverUrl);
+                    // This will overwrite any previous entry if the digest is the same.
+                    // However, that's not a problem, since we're intentionally deduping
+                    // images with the same payload anyway. It doesn't matter which digest
+                    // we use in the end, all of those digests return the same data.
+                    dataToOriginal.set(digest, coverUrl);
+                    return trackCovers.map((cover) => {
+                        return {
+                            ...cover,
+                            url: digest,
```

Comment:
This assignment is a bit confusing, a comment to indicate that this happens purely to satisfy the requirements of `#groupIdenticalImages()` would clarify this.

Reply:
Refactored the implementation and sprinkled in some more comments. It should be more elegant and clear now.